pkg/prometheus/operator.go

```diff
@@ -1937,35 +1646,58 @@ func (c *Operator) selectServiceMonitors(p *monitoringv1.Prometheus) (map[string
 		cache.ListAllByNamespace(c.smonInf.GetIndexer(), ns, servMonSelector, func(obj interface{}) {
 			k, ok := c.keyFunc(obj)
 			if ok {
-				res[k] = obj.(*monitoringv1.ServiceMonitor)
+				serviceMonitors[k] = obj.(*monitoringv1.ServiceMonitor)
 			}
 		})
 	}
 
-	// If denied by Prometheus spec, filter out all service monitors that access
-	// the file system.
-	if p.Spec.ArbitraryFSAccessThroughSMs.Deny {
-		for namespaceAndName, sm := range res {
-			for _, endpoint := range sm.Spec.Endpoints {
-				if err := testForArbitraryFSAccess(endpoint); err != nil {
-					delete(res, namespaceAndName)
-					level.Warn(c.logger).Log(
-						"msg", "skipping servicemonitor",
-						"error", err.Error(),
-						"servicemonitor", namespaceAndName,
-						"namespace", p.Namespace,
-						"prometheus", p.Name,
-					)
+	res := make(map[string]*monitoringv1.ServiceMonitor, len(serviceMonitors))
+	for namespaceAndName, sm := range serviceMonitors {
+		var err error
+
+		for i, endpoint := range sm.Spec.Endpoints {
+			// If denied by Prometheus spec, filter out all service monitors that access
+			// the file system.
+			if p.Spec.ArbitraryFSAccessThroughSMs.Deny {
+				if err = testForArbitraryFSAccess(endpoint); err != nil {
+					break
 				}
 			}
+
+			smKey := fmt.Sprintf("serviceMonitor/%s/%s/%d", sm.GetNamespace(), sm.GetName(), i)
+
+			if err = store.addBearerToken(sm.GetNamespace(), endpoint.BearerTokenSecret, smKey); err != nil {
+				break
+			}
+
+			if err = store.addBasicAuth(sm.GetNamespace(), endpoint.BasicAuth, smKey); err != nil {
+				break
+			}
+
+			if err = store.addTLSConfig(sm.GetNamespace(), endpoint.TLSConfig); err != nil {
+				break
+			}
+		}
+
+		if err != nil {
+			level.Warn(c.logger).Log(
+				"msg", "skipping servicemonitor",
+				"error", err.Error(),
+				"servicemonitor", namespaceAndName,
+				"namespace", p.Namespace,
+				"prometheus", p.Name,
+			)
```

Comment:
do we want to add a new counting metric too? I believe this would be valuable to track.